frontend/src/Frontend/Wallet.hs

```diff
@@ -224,31 +225,49 @@ getBalances
     )
-  => model -> Event t AccountStorage -> m (Event t [(NetworkName, Some AccountRef, Maybe AccountBalance)])
-getBalances model accStore = performEventAsync $ flip push accStore $ \networkAccounts -> do
+  => model -> Event t (KeyStorage key, AccountStorage) -> m (Event t [(NetworkName, Some AccountRef, Maybe AccountBalance)])
+getBalances model accStore = performEventAsync $ flip push accStore $ \(keys, networkAccounts) -> do
   nodes <- fmap rights $ sample $ current $ model ^. network_selectedNodes
   net <- sample $ current $ model ^. network_selectedNetwork
-  pure $ mkRequests nodes net <$> Map.lookup net (unAccountStorage networkAccounts)
+  pure . Just $ mkRequests nodes net keys (Map.lookup net (unAccountStorage networkAccounts))
   where
-    mkRequests nodes net accounts cb = do
+    allChains = ChainId . tshow <$> ([0..9] :: [Int])
+    allChainsLen = length allChains
+    onAllChains = MonoidalMap.fromList . zip allChains . replicate allChainsLen
+
+    mkRequests nodes net keys mAccounts cb = do
       -- Transform the vanity accounts structure into a map from chain ID to
-      -- set of account names
-      let chainsToAccounts = flip foldAccounts accounts $ \case
-            AccountRef_Vanity (AccountName name) chain :=> Identity acc
-              | _accountInfo_hidden $ _vanityAccount_info acc -> mempty
-              | otherwise -> MonoidalMap.singleton chain $ Set.singleton name
-            AccountRef_NonVanity pk chain :=> Identity acc
-              | _accountInfo_hidden $ _nonVanityAccount_info acc -> mempty
-              | otherwise -> MonoidalMap.singleton chain $ Set.singleton $ keyToText pk
-          code = renderCompactText . accountBalanceObject . Set.toList
-          pm chain = Pact.PublicMeta
-            { Pact._pmChainId = chain
-            , Pact._pmSender = "chainweaver"
-            , Pact._pmGasLimit = defaultTransactionGasLimit
-            , Pact._pmGasPrice = defaultTransactionGasPrice
-            , Pact._pmTTL = 3600
-            , Pact._pmCreationTime = 0
-            }
+      -- set of account names. We grab balances even for keys on all chains, but
+      -- stick to only the chains that we know for vanity accounts otherwise we
+      -- could easily prevent the user from being able to create accounts that
+      -- they may wish to create (not to mention spamming extra vanity accounts
+      -- makes the accounts page look ugly).
+      --
+      -- We spam extra balance requests here for keys rather than just adding the
+      -- accounts on generateKey because if we just did it on create then we'd
+      -- get the user stuck if they ever added in a new network.
+      let
+        chainsToAccounts = fold
+            [ foldMap
+              (\(an, m) ->
+                foldMap (\cId ->
+                  MonoidalMap.singleton cId (Set.singleton (unAccountName an)))
+                (Map.keys m))
+              (mAccounts ^.. _Just . accounts_vanity . to Map.toList . traverse)
```

Comment:
How about 
``` haskell
fold
  [ flip foldMap mAccounts $ foldAccounts $ \case
    AccountRef_NonVanity _ _ :=> _ -> mempty
    AccountRef_Vanity an cId :=> _ -> MonoidalMap.singleton cId (Set.singleton (unAccountName an))
  ,
```

Reply:
Fold accounts didn't work here anymore, but I forget why. I'll try to remember. :)

Reply:
It typechecks but can be much less efficient since it visits all the non-vanity accounts for what's basically a no-op, which could probably be fixed with 
```haskell
flip foldMap mAccounts $ \a -> flip foldAccounts (a & accounts_vanity .~ mempty) $ \case
```
but at that point the readability gain is somewhat lost.

Let's leave it as is. If this usecase starts showing up a lot we can just break `foldAccounts` up into vanity and nonvanity.